CompositeException issue

I'm using RxJava on Android and I've encountered issues with CompositeException. Because it collects throwables as `Collection<Throwable> errors` the exceptions are not logged to the logcat output if there is an unhandled Exception in onError callback. 

Here is a sample output:

```
2739:E/AndroidRuntime(14639): FATAL EXCEPTION: main
2740:E/AndroidRuntime(14639): rx.exceptions.OnErrorFailedException: Error occurred when trying to propagate error to Observer.onError
2741:E/AndroidRuntime(14639):   at rx.observers.SafeSubscriber._onError(SafeSubscriber.java:182)
2742:E/AndroidRuntime(14639):   at rx.observers.SafeSubscriber.onError(SafeSubscriber.java:103)
2743:E/AndroidRuntime(14639):   at rx.internal.operators.NotificationLite.accept(NotificationLite.java:144)
2744:E/AndroidRuntime(14639):   at rx.internal.operators.OperatorObserveOn$ObserveOnSubscriber.pollQueue(OperatorObserveOn.java:139)
2745:E/AndroidRuntime(14639):   at rx.internal.operators.OperatorObserveOn$ObserveOnSubscriber.access$000(OperatorObserveOn.java:61)
2746:E/AndroidRuntime(14639):   at rx.internal.operators.OperatorObserveOn$ObserveOnSubscriber$1.call(OperatorObserveOn.java:121)
2747:E/AndroidRuntime(14639):   at rx.internal.schedulers.ScheduledAction.run(ScheduledAction.java:43)
2748:E/AndroidRuntime(14639):   at android.os.Handler.handleCallback(Handler.java:730)
2749:E/AndroidRuntime(14639):   at android.os.Handler.dispatchMessage(Handler.java:92)
2750:E/AndroidRuntime(14639):   at android.os.Looper.loop(Looper.java:137)
2751:E/AndroidRuntime(14639):   at android.app.ActivityThread.main(ActivityThread.java:5103)
2752:E/AndroidRuntime(14639):   at java.lang.reflect.Method.invokeNative(Native Method)
2753:E/AndroidRuntime(14639):   at java.lang.reflect.Method.invoke(Method.java:525)
2754:E/AndroidRuntime(14639):   at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:737)
2755:E/AndroidRuntime(14639):   at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553)
2756:E/AndroidRuntime(14639):   at dalvik.system.NativeStart.main(Native Method)
2757:E/AndroidRuntime(14639): Caused by: rx.exceptions.CompositeException: 2 exceptions occurred. See them in causal chain below.
2758:E/AndroidRuntime(14639):   ... 16 more
```

That's it. No logs about real cause. I can wrap my onError code in try catch block and log the real cause, but I think this can be handled better by RxJava. What do you guys think?

--- a/rxjava-core/src/main/java/rx/exceptions/CompositeException.java
+++ b/rxjava-core/src/main/java/rx/exceptions/CompositeException.java
@@ -38,9 +38,9 @@ public final class CompositeException extends RuntimeException {
     public CompositeException(String messagePrefix, Collection<Throwable> errors) {
         List<Throwable> _exceptions = new ArrayList<Throwable>();
         CompositeExceptionCausalChain _cause = new CompositeExceptionCausalChain();
-        int count = 0;
+        int count = errors.size();
+        errors = removeDuplicatedCauses(errors);
         for (Throwable e : errors) {
-            count++;
             attachCallingThreadStack(_cause, e);
             _exceptions.add(e);
         }
@@ -71,6 +71,29 @@ public final class CompositeException extends RuntimeException {
     @Override
     public synchronized Throwable getCause() {
         return cause;
+    }
+
+    private Collection<Throwable> removeDuplicatedCauses(Collection<Throwable> errors) {
+        Set<Throwable> duplicated = new HashSet<Throwable>();
+        for (Throwable cause : errors) {
+            for (Throwable error : errors) {
+                if(cause == error || duplicated.contains(error)) {
+                    continue;
+                }
+                while (error.getCause() != null) {
+                    error = error.getCause();
+                    if (error == cause) {
+                        duplicated.add(cause);
+                        break;
+                    }
+                }
+            }
+        }
+        if (!duplicated.isEmpty()) {
+            errors = new ArrayList<Throwable>(errors);
+            errors.removeAll(duplicated);
+        }
+        return errors;
     }
 
     @SuppressWarnings("unused")

--- a/rxjava-core/src/test/java/rx/exceptions/CompositeExceptionTest.java
+++ b/rxjava-core/src/test/java/rx/exceptions/CompositeExceptionTest.java
@@ -54,6 +54,16 @@ public class CompositeExceptionTest {
     }
 
     @Test(timeout = 1000)
+    public void testOneIsCauseOfAnother() {
+        Throwable rootCause = new Throwable("RootCause");
+        Throwable throwable = new Throwable("1", rootCause);
+        CompositeException ce = new CompositeException("One is the cause of another",
+                Arrays.asList(rootCause, throwable));
+
+        assertEquals(1, ce.getExceptions().size());
+    }
+
+    @Test(timeout = 1000)
     public void testAttachCallingThreadStackParentThenChild() {
         CompositeException.attachCallingThreadStack(ex1, ex2);
         assertEquals("Ex2", ex1.getCause().getMessage());

